gem/unconcatenate.py

```diff
@@ -0,0 +1,192 @@
+"""Utility functions for decomposing Concatenate nodes."""
+
+from __future__ import absolute_import, print_function, division
+from six.moves import range, zip
+
+from itertools import chain
+
+import numpy
+
+from gem.node import Memoizer, reuse_if_untouched
+from gem.gem import (ComponentTensor, Concatenate, FlexiblyIndexed,
+                     Index, Indexed, Node, reshape, view)
+from gem.optimise import remove_componenttensors
+
+
+__all__ = ['unconcatenate']
+
+
+def find_group(expressions):
+    """Finds a full set of indexed Concatenate nodes with the same
+    free index, if any such node exists.
+
+    Pre-condition: ComponentTensor nodes surrounding Concatenate nodes
+    must be removed.
+
+    :arg expressions: a multi-root GEM expression DAG
+    :returns: a list of GEM nodes, or None
+    """
+    free_indices = set().union(chain(*[e.free_indices for e in expressions]))
+
+    # Result variables
+    index = None
+    nodes = []
+
+    # Sui generis pre-order traversal so that we can avoid going
+    # unnecessarily deep in the DAG.
+    seen = set()
+    lifo = []
+    for root in expressions:
+        if root not in seen:
+            seen.add(root)
+            lifo.append(root)
+
+    while lifo:
+        node = lifo.pop()
+        if not free_indices.intersection(node.free_indices):
+            continue
+
+        if isinstance(node, Indexed):
+            child, = node.children
+            if isinstance(child, Concatenate):
+                i, = node.multiindex
+                assert i in free_indices
+                if (index or i) == i:
```

Comment:
OK, so if this is the first `Concatenate`, collect it, otherwise only collect it if its free indices match.  So this doesn't produce all `Concatenate` nodes, just a subset with matching free indices (but not the list of such subsets).